Enable API documentation for `std::os::wasi`.

sunfishcode opened the pull request:
This adds API documentation support for `std::os::wasi` modeled after
how `std::os::unix` works, so that WASI can be documented [here] along
with the other platforms.

[here]: https://doc.rust-lang.org/stable/std/os/index.html

Two changes of particular interest:

 - This changes the `AsRawFd` for `io::Stdin` for WASI to return
   `libc::STDIN_FILENO` instead of `sys::stdio::Stdin.as_raw_fd()` (and
   similar for `Stdout` and `Stderr`), which matches how the `unix`
   version works. `STDIN_FILENO` etc. may not always be explicitly
   reserved at the WASI level, but as long as we have Rust's `std` and
   `libc`, I think it's reasonable to guarantee that we'll always use
   `libc::STDIN_FILENO` for stdin.

 - This duplicates the `osstr2str` utility function, rather than
   trying to share it across all the configurations that need it.

r? @alexcrichton

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 643d2ce3e2de481593bc4aaee81f1df361680864 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 643d2ce3e2de481593bc4aaee81f1df361680864 -->
<!-- homu: {"type":"Approved","sha":"643d2ce3e2de481593bc4aaee81f1df361680864","approver":"alexcrichton"} -->

camelid commented on library/std/src/os/mod.rs:
```diff
@@ -17,6 +17,9 @@ cfg_if::cfg_if! {
         #[stable(feature = "rust1", since = "1.0.0")]
         pub use crate::sys::windows_ext as windows;
 
+        #[cfg(not(target_os = "wasi"))]
+        pub use crate::sys::wasi_ext as wasi;
```
Does this not need a `stable`/`unstable` attribute?

sunfishcode replied:
I don't know. The existing `linux` module doesn't have one, so I assumed WASI didn't either, but I'd be happy to add one otherwise.

camelid replied:
Maybe it's not necessary, but I saw it on the `windows` and `unix` ones, so I thought I'd check.

CDirkx replied:
The linux module has a stability attribute, just inside the module itself 

https://github.com/rust-lang/rust/blob/019610754363d1d92a8d0f364d2c0909d6f53dfd/library/std/src/os/linux/mod.rs#L3

I would think this requires a stability attribute, but I'm not sure because the tests didn't complain.

bors commented:
:umbrella: The latest upstream changes (presumably #82430) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1961534522) [(plain)](https://github.com/rust-lang/rust/commit/5c46468f5fc0943a6cbaca089a17b082629fda2f/checks/1961534522/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
######################################################################## 100.0%
extracting /checkout/obj/build/cache/2021-01-28/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
    Updating crates.io index
    Updating git repository `https://github.com/rust-lang/cargo`
error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, use the --offline flag.
Build completed unsuccessfully in 0:00:43
make: *** [prepare] Error 1
Makefile:64: recipe for target 'prepare' failed
Command failed. Attempt 2/5:
Command failed. Attempt 2/5:
    Updating crates.io index
    Updating git repository `https://github.com/rust-lang/cargo`
error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, use the --offline flag.
Build completed unsuccessfully in 0:00:00
make: *** [prepare] Error 1
Makefile:64: recipe for target 'prepare' failed
Command failed. Attempt 3/5:
Command failed. Attempt 3/5:
    Updating crates.io index
    Updating git repository `https://github.com/rust-lang/cargo`
error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, use the --offline flag.
Build completed unsuccessfully in 0:00:00
make: *** [prepare] Error 1
Makefile:64: recipe for target 'prepare' failed
Command failed. Attempt 4/5:
Command failed. Attempt 4/5:
    Updating crates.io index
    Updating git repository `https://github.com/rust-lang/cargo`
error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, use the --offline flag.
Build completed unsuccessfully in 0:00:02
Makefile:64: recipe for target 'prepare' failed
make: *** [prepare] Error 1
Command failed. Attempt 5/5:
Command failed. Attempt 5/5:
    Updating crates.io index
    Updating git repository `https://github.com/rust-lang/cargo`
error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, use the --offline flag.
Build completed unsuccessfully in 0:00:01
make: *** [prepare] Error 1
Makefile:64: recipe for target 'prepare' failed
The command has failed after 5 attempts.
```

</details>

alexcrichton commented:
@bors: r+ delegate+

bors commented:
:pushpin: Commit 132ec261b0d215f77b71eeee3d814cd304fc1a29 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 132ec261b0d215f77b71eeee3d814cd304fc1a29 -->
<!-- homu: {"type":"Approved","sha":"132ec261b0d215f77b71eeee3d814cd304fc1a29","approver":"alexcrichton"} -->

bors commented:
:v: @sunfishcode can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"alexcrichton","delegate":"sunfishcode"} -->

CDirkx commented on library/std/src/sys/mod.rs:
```diff
@@ -106,3 +106,25 @@ cfg_if::cfg_if! {
+cfg_if::cfg_if! {
+    if #[cfg(target_os = "wasi")] {
+        // On WASI we'll document what's already available
+        #[stable(feature = "rust1", since = "1.0.0")]
```
This should probably be updated

CDirkx commented:
Currently the documentation does not build on all platforms:

https://github.com/rust-lang/rust/blob/fe1bf8e05c39bdcc73fc09e246b7209444e389bc/library/std/src/sys/wasi/ext/mod.rs#L11-L24

All these should use `super::` instead of `crate::sys::ext` because with this change, when documenting on windows for example, `sys::ext` will be the windows `ext`, not the wasi one.

CDirkx commented:
![Capture](https://user-images.githubusercontent.com/11176562/108910240-e1420580-7625-11eb-9a56-970f8beb87f8.PNG)

Rustdoc will also need to be updated, but that can be a followup.

JohnTitor commented:
Failed in rollup: https://github.com/rust-lang/rust/pull/82472#issuecomment-785124567
@bors r-

sunfishcode commented:
Thanks! I've now fixed the `crate::sys::ext` paths, added the `cfg(not(doc))` to the main `wasi` module, and added "wasi" to the "main module" comment.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 7d5242a03a9e30fe880ff1af83a3ad7df5e62159 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 7d5242a03a9e30fe880ff1af83a3ad7df5e62159 -->
<!-- homu: {"type":"Approved","sha":"7d5242a03a9e30fe880ff1af83a3ad7df5e62159","approver":"alexcrichton"} -->
